### ISSUE TITLE
🐛 [SignalK] Connecting to ttyUSB*

### ISSUE THREAD

**Issue by MrHobsons** [alexbelgium/hassio-addons#1450]
### Description

Not sure if it is a bug or not but I'm not able to get a connection via ttyUSB0 or ttyUSB0 under Connections in SignalK

![image](https://github.com/alexbelgium/hassio-addons/assets/154426671/704e60f7-b61f-4a9d-8282-56443d87bbf3)
![image](https://github.com/alexbelgium/hassio-addons/assets/154426671/5be2c37e-c3d1-4b56-9d1a-48ac3dfb500b)

### Reproduction steps

```bash
1. Go to 'SignalK'
2. Click on 'Data Connections'
```

### Addon Logs

```bash
Registering new address record for ::1 on lo.*.
Registering new address record for 127.0.0.1 on lo.IPv4.
Server startup complete. Host name is db21ed7f-signalk.local. Local service cookie is 1282224705.
*** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/blog/projects/avahi-compat.html>
*** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/blog/projects/avahi-compat.html>
signalk-server running at 0.0.0.0:3000

Error: Permission denied, cannot open /dev/ttyUSB1
GET / 302 44.263 ms - 58
GET /admin/ 200 29.302 ms - 562
GET /admin/main.js 304 16.176 ms - -
GET /@signalk/vesselpositions/remoteEntry.js 304 51.468 ms - -
GET /admin/208.js 304 20.925 ms - -
GET /admin/41.js 304 15.591 ms - -
GET /admin/144.js 304 8.884 ms - -
GET /admin/259.js 304 7.148 ms - -
GET /skServer/plugins 304 58.821 ms - -
GET /skServer/loginStatus 200 12.505 ms - 204
```

### OS

HAos

**Comment by alexbelgium** [comment]
Hi, I need to map access to it. Any other devices I should add at the same time?

**Comment by alexbelgium** [comment]
OK done

**Comment by MrHobsons** [comment]
Hi Alexandre,

Ok, I have tried again but no luck, error  still the same:
Jun 29 23:42:29 Error: Permission denied, cannot open /dev/ttyUSB1

Any ideas?




reagrds /anders



________________________________

**Comment by alexbelgium** [comment]
Hi, perhaps I need to adapt apparmor permissions in addition to the config options, I'll try that later today

**Comment by MrHobsons** [comment]
Hi,

Tried the latest version but no success, still same error message




Mvh /anders



________________________________

**Comment by alexbelgium** [comment]
I'm sorry it's the only configurable options linked to ttyusb that HA proposes I am not sure how to help...

**Comment by MrHobsons** [comment]
Hi,

ok,  I'm also using the node-red addon and from that I can access ttyUSB0 and ttyUSB1
but I really need it to work with the signal k addon

Could you look at the node-red addon and see if you can get some info regarding this issue?

https://github.com/hassio-addons/addon-node-red
[https://opengraph.githubassets.com/af5ec5bb7f72a58de7f5dc4b3d61354d8a80d1c7606c3afa12703874f4cd020d/hassio-addons/addon-node-red]<https://github.com/hassio-addons/addon-node-red>
GitHub - hassio-addons/addon-node-red: Node-RED - Home Assistant Community Add-ons<https://github.com/hassio-addons/addon-node-red>
Node-RED - Home Assistant Community Add-ons. Contribute to hassio-addons/addon-node-red development by creating an account on GitHub.
github.com


regards /anders



________________________________

**Comment by alexbelgium** [comment]
This addon has TONS of permissions added... I can try to add all of those but it will be preferable to remove it after to avoid creating security risks.

If you are ok I could duplicate the addon on my test repo and add the permissions there

**Comment by MrHobsons** [comment]
Hi,

ok I understand.

It would be great if you could do that, but I’m
not sure how I can update to that version.

regards /anders

Skickat från Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Alexandre ***@***.***>
Skickat: Monday, July 1, 2024 4:30:10 PM
Till: alexbelgium/hassio-addons ***@***.***>
Kopia: MrHobsons ***@***.***>; Author ***@***.***>
Ämne: Re: [alexbelgium/hassio-addons] 🐛 [SignalK] Connecting to ttyUSB* (Issue #1450)


This addon has TONS of permissions added... I can try to add all of those but it will be preferable to remove it after to avoid creating security risks.

If you are ok I could duplicate the addon on my test repo and add the permissions there

**Comment by alexbelgium** [comment]
Hi,

Can you please install the repository https://github.com/alexbelgium/hassio-addons-test ? It will contain an addon named "Signalk Server Test" that has all possible permissions found in the nodered addon. We'll adapt depending on that ;-)

Thanks for your help!

**Comment by alexbelgium** [comment]
I'll also push an update to the main repo version with a potential fix from https://github.com/esp8266/source-code-examples/issues/26

**Comment by MrHobsons** [comment]
Hi,

Thanks, I will try this tomorrow and get back to you.

regards /anders

Skickat från Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Alexandre ***@***.***>
Skickat: Tuesday, July 2, 2024 7:10:45 PM
Till: alexbelgium/hassio-addons ***@***.***>
Kopia: MrHobsons ***@***.***>; Author ***@***.***>
Ämne: Re: [alexbelgium/hassio-addons] 🐛 [SignalK] Connecting to ttyUSB* (Issue #1450)


Hi,

Can you please install the repository https://github.com/alexbelgium/hassio-addons-test ? It will contain an addon named "Signalk Server Test" that has all possible permissions found in the nodered addon. We'll adapt depending on that ;-)

Thanks for your help!

**Comment by MrHobsons** [comment]
Hi,

Tried the "Signalk Server Test" but unfortunately same error.


 Regards /anders



________________________________

**Comment by MrHobsons** [comment]
Hi,

Di you also put this in Test?




Regards /anders



________________________________

**Comment by alexbelgium** [comment]
Hi, "Test" was only with the same config.yaml options than the node red addon (which therefore are confirmed not to be the solution), however the version on the main repo has a specific permission added that should solve the issue according to what I read on some other projects

**Comment by MrHobsons** [comment]
Hi, 
So version 2.8.3-4 has has a specific permission added that should solve the issue ?

**Comment by alexbelgium** [comment]
That's indeed what I wrote, is it working?

**Comment by MrHobsons** [comment]
Hi,

No, unfortunately it’s not working.
Same error permission denied.

Regards  /anders

Skickat från Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Alexandre ***@***.***>
Skickat: Wednesday, July 10, 2024 11:37:23 AM
Till: alexbelgium/hassio-addons ***@***.***>
Kopia: MrHobsons ***@***.***>; Author ***@***.***>
Ämne: Re: [alexbelgium/hassio-addons] 🐛 [SignalK] Connecting to ttyUSB* (Issue #1450)


That's indeed what I wrote, is it working?

**Comment by alexbelgium** [comment]
Hi, mmh that is unfortunate as this was the only solution I found on github for other projects corresponding to this issue... I would have to see if I have a mean to replicate this type of device on my system to be able to really find a solution as curently I can only rely on blind tries based on other's solutions

**Comment by alexbelgium** [comment]
I couldn't replicate but I have an idea. If you have the addons File Editor or Filebrowser, you can find the file `/addon_configs/db21ed7f_signalk/signalk.sh` . Everything you place in this file will be executed at addon start.

So if you add in this file the code below, it should print you which are the permissions required to access the device. In particular in terms of usergroup.
```
#!/usr/bin/with-contenv bashio
# shellcheck shell=bash

#################
# CODE INJECTOR #
#################

# Any commands written in this bash script will be executed at addon start
# See guide here : https://github.com/alexbelgium/hassio-addons/wiki/Add%E2%80%90ons-feature-:-customisation
log="$(ls -l /dev/ttyUSB0)"
bashio::log.warning "This is the required permissions : $log"
```

**Comment by MrHobsons** [comment]
Hi,

This is the result:

/etc/cont-init.d/01-custom_script.sh: executing
Execute /addon_configs/db21ed7f-signalk/signalk.sh if existing
Wiki here : github.com/alexbelgium/hassio-addons/wiki/Add-ons-feature-:-customisation
... script found, executing
[05:32:01] WARNING: This is the required permissions : crw-rw---- 1 root 18 188, 0 Jul 11 05:22 /dev/ttyUSB0
[05:32:01] WARNING: This is the required permissions : crw-rw---- 1 root 18 188, 1 Jul 11 05:22 /dev/ttyUSB1
/etc/cont-init.d/99-run.sh: executing
[05:32:01] INFO: Starting application


Regards /anders



________________________________

**Comment by alexbelgium** [comment]
That's great thanks! So write read/access is granted to root and usergroup, with usergroup being "18". I detected an error in the code that I should have set the permissions. I'll push a new version now, it should take around 15 minutes to build (until that time you would get a manifest unknown error in homeassistant if you try to install it). Thanks for your patience hope this is will be good this time !

**Comment by MrHobsons** [comment]
Hi,

Perfect, thanks.

I will try this and let you know



Regards/anders



________________________________

**Comment by MrHobsons** [comment]
Hi,

Did the update and now I have another error and it never starts

/etc/cont-init.d/01-custom_script.sh: executing
Execute /addon_configs/db21ed7f-signalk/signalk.sh if existing
Wiki here : github.com/alexbelgium/hassio-addons/wiki/Add-ons-feature-:-customisation
/etc/cont-init.d/99-run.sh: executing
usermod: group 'UNKNOWN' does not exist



Regards /anders



________________________________

**Comment by alexbelgium** [comment]
Ah - that's a good sign! I'll update the script and push a new version thanks

**Comment by MrHobsons** [comment]
Hi,

Unfortunately I'm getting Permission error again

Network interface enumeration completed.
Registering new address record for 172.30.33.6 on eth0.IPv4.
Registering new address record for ::1 on lo.*.
Registering new address record for 127.0.0.1 on lo.IPv4.
Server startup complete. Host name is db21ed7f-signalk.local. Local service cookie is 409187000.
signalk-server running at 0.0.0.0:3000

Error: Permission denied, cannot open /dev/ttyUSB1





Mvh /anders



________________________________

**Comment by alexbelgium** [comment]
Sorry for that. Then most likely I'll have to replicate your system on mine to troubleshoot, alas I will be on holidays for 2 weeks so I won't have time to check again probably soon. Except if you also use the .sh method to inject other code to solve the issue...

**Comment by alexbelgium** [comment]
Hi, the key thing is that signalk is ran under the user `node` which
doesn't seem to have access to the device as only root and it's subgroup
have access to it.

We can't increase /dev/ttyUSB permissions as HA mounts it read only so the
only way I was thinking (the last version) was therefore to add node to the
root group but it didn't work. I'll have to think what more we can try...

**Comment by slundell** [comment]
In addition to the USB devices many will need access to the socketcan device /dev/can0 and maybe /dev/can1 to be able to read NMEA2000 data.
These would also need to be initialised with these commands

ip link set can0 up type can bitrate 250000
ifconfig can0 txqueuelen 65536

neither ip nor ifconfig is accessable in the container, but these can be installed by adding the package iproute2. I do not think can-utils will be needed.

**Comment by slundell** [comment]
I have messed around, trying to get a decent dev env going. I am tunneling to my boat where I have a rpi cm4 running with an mp2515 connected to can0.  I suspect that the same kind of access issue that is in play with USB will be in play with can, hence why I'm hijacking this issue. 

What would be the best way to setup a minimal env on the boat rpi to be able to make local changes to the addon?

**Comment by alexbelgium** [comment]
Hi, to make local changes shrub the add-on the best is to use my infection script described above.  With it, you can even prevent the normal add-on launch and instead launch it with a custom command. However, it is not possible to modify the add-on configuration (such as permissions) with that, as HA reads the config.json embedded at build time.

I've just thought that to simplify things, I'll push a new build with the "full access" option. From the main option you can decide to give admin access to the add-on and it should have all permissions. If this work, then we can use the injection script to execute commands with admin access disabled. Or we just keep full access, but it is usually not preferable

**Comment by MrHobsons** [comment]
Hi

I think the "full access" would be nice while we
figure out how to solve this issue.

Regards /anders

Skickat från Outlook för iOS<https://aka.ms/o0ukef>
________________________________
Från: Alexandre ***@***.***>
Skickat: Sunday, July 14, 2024 9:09:56 AM
Till: alexbelgium/hassio-addons ***@***.***>
Kopia: MrHobsons ***@***.***>; Author ***@***.***>
Ämne: Re: [alexbelgium/hassio-addons] 🐛 [SignalK] Connecting to ttyUSB* (Issue #1450)


Hi, to make local changes shrub the add-on the best is to use my infection script described above. With it, you can even prevent the normal add-on launch and instead launch it with a custom command. However, it is not possible to modify the add-on configuration (such as permissions) with that, as HA reads the config.json embedded at build time.

I've just thought that to simplify things, I'll push a new build with the "full access" option. From the main option you can decide to give admin access to the add-on and it should have all permissions. If this work, then we can use the injection script to execute commands with admin access disabled. Or we just keep full access, but it is usually not preferable

**Comment by slundell** [comment]
It seems like can0 is not exposed to the signalk container. If i run "ip link" to see the network connections in the injection script there is only lo and eth0. If i do the same in the ssh terminal addon, i get a bunch of interfaces, including can0.

**Comment by frostworx** [comment]
I don't use the addon (in fact it is not even installed), but you might want to know that 

hassio_supervisor logs complain with

`2024-07-15 19:21:34.805 WARNING (SyncWorker_7) [supervisor.addons.validate] Add-on have full device access, and selective device access in the configuration. Please report this to the maintainer of Signalk Server`

_(I've read above that this was activated on purpose, but you possibly haven't noticed that this warning pops up generally)_

I personally do not care, but wanted to find out what a "Signalk Server" and therefore landed here :)

**Comment by alexbelgium** [comment]
> I don't use the addon (in fact it is not even installed), but you might want to know that
> 
> hassio_supervisor logs complain with
> 
> `2024-07-15 19:21:34.805 WARNING (SyncWorker_7) [supervisor.addons.validate] Add-on have full device access, and selective device access in the configuration. Please report this to the maintainer of Signalk Server`
> 
> _(I've read above that this was activated on purpose, but you possibly haven't noticed that this warning pops up generally)_
> 
> I personally do not care, but wanted to find out what a "Signalk Server" and therefore landed here :)

Thanks that's because I enabled full access in latest version but kept the partial access to revert once it is working ;)

**Comment by MrHobsons** [comment]
Hi,

I have np tried the "full access" but it did not solve the problem with permission denied om /dev/ttyUSB1

regards /anders



________________________________

**Comment by tingles2233** [comment]
Been following along this curious to see where it ended up as I will also be needing usb access to signalk in the near future.

**Comment by tingles2233** [comment]
Also wanted to add I am willing/able to test anything required.

**Comment by paalkr** [comment]
I'm also interested in can0 access. I plan to power my HA RPi using the Pican-M hat connected to the onboard NMEA2000 bus. The Pican-M board exposed the n2k interface on can0.

**Comment by tingles2233** [comment]
Was doing some research on this I wonder if this is a similar issue, which came up with VenusOS (Victron Controller) which also has the ability to run Signalk in a container.. That being said it may not be relevant.

https://communityarchive.victronenergy.com/questions/138314/ngt-1-pi-and-venus.html

**Comment by paalkr** [comment]
This commit https://github.com/alexbelgium/hassio-addons/commit/b07bdce9d8bb1dcded710c0b182d6dcaf95703f4 apparently adds support for /dev/can0

**Comment by tingles2233** [comment]
Interesting. 

In the past hour I setup a fresh Hass installation on a RP5 with a USB Device (NGX-1-USB) to map to Signalk. 

Still receiving the error presented before. "[Error: Error: Permission denied, cannot open /dev/ttyUSB0]" 

Core: 2024.9.2
Supervisor: 2024.09.1
Operating System: 13.1
Frontend: 20240909.1
Signal K Addon Version: 2.11.0 

Trying to see if there are any other work arounds I can perform to make it work in this environment and will update if anything interesting is found.

**Comment by paalkr** [comment]
This is from a rather old issue, not sure it's even relevant
https://github.com/home-assistant/supervisor/issues/2690#issuecomment-809321072

**Comment by paalkr** [comment]
This is not my area of expertise, but maybe the /dev/can0 will work in the current state because it's a serial interface and not a tty interface? Seems like HA somehow limits access to tty in the container addons. However, there are plenty of addons, like zigbee2mqtt and zwave2mqtt that allows access to tty devices inside the containers.

BTW, the zigbee2mqtt addon does not have an apparmor.txt file. The signalk addon does have that, but /dev/can0 is not mentioned. https://github.com/alexbelgium/hassio-addons/blob/master/signalk/apparmor.txt

PS, I dont have access to my pican-m board at home right now, as it's currently in use on another RPi in my powerboat.

**Comment by tingles2233** [comment]
Bump to keep this issue current. Any updates?

**Comment by tingles2233** [comment]
Following up on this as I recently had to move my signalk instance off of HASS because of the lack of TTYUSB support and running a separate rpi just for signalk. Would love to move it back as it is the only part that was not working. 

Unfortunately I am not exactly sure what is not mapped correctly within the addon, but there are other addons that use the local usb port and devices attached. 

Willing to test anything required.

**Comment by paalkr** [comment]
@tingles2233 , I ended up doing the exact same thing, running SignalK on a separate Pi4 in my boat. Data from SignalK is exported to HA through mqtt, and it's actually working great. Running everything on the same Pi would have been nice to reduce the number of devices to integrate, but I think it will be hard to get the Pican-M HAT to function properly on Hass OS. Even if I managed to get the Pican-M HAT including the binary dependencies to function in Hass OS, it's still the issue of exposing the can0 device to the SignalK addon that I believe will be hard. So, I'm quite happy with running two separate devices. The SignalK Pi services AIS data to N2K network onboard to use in my Orca and Garmin nav stations. I think running SingalK on a separate device is a good decision for redundancy.

**Comment by alexbelgium** [comment]
Hi, I've tried pushing a new version with uart and removal of apparmor just to try, I don't see what other difference zigbee2mqtt has...

**Comment by alexbelgium** [comment]
> This is from a rather old issue, not sure it's even relevant [home-assistant/supervisor#2690 (comment)](https://github.com/home-assistant/supervisor/issues/2690#issuecomment-809321072)

Thanks very much! It confirmed that `uart: true` is needed for tty so I implemented it in my last build that should hopefully work!

**Comment by tingles2233** [comment]
Awesome thanks Alex I do appreciate your help with this!

I`ll test it out over the next 48-72 hours and update this issue.

**Comment by alistairg** [comment]
I'm having this same issue with the latest release. My device is ultimaterly /dev/ttyACM0, and I get permission denied for it.

Any ideas?

**Comment by alexbelgium** [comment]
Thanks! Added in a new version

**Comment by alistairg** [comment]
> Thanks! Added in a new version

Thankyou! Sadly same error:

`Error: Permission denied, cannot open /dev/ttyACM0`

**Comment by alexbelgium** [comment]
Argh thanks. I'll see a bit later to add back the full access toggle in case it helps

**Comment by alexbelgium** [comment]
I've disabled app armor does it work?

**Comment by alistairg** [comment]
> I've disabled app armor does it work?

Sadly not :(

**Comment by alistairg** [comment]
> I've disabled app armor does it work?

Sadly not :(

**Comment by alexbelgium** [comment]
And now? I've removed all possible protections :) please try with the full access flag enabled or disabled

**Comment by MrHobsons** [comment]
Hi,

I have tried but no success.


 Regards /anders



________________________________

**Comment by alexbelgium** [comment]
Well then I guess then it's an issue with HA itself, there is nothing I can do more from my side...

Merry Christmas anyway!

**Comment by tingles2233** [comment]
Hey team, 

By chance if we explicitly found and listed then selected a device similar to how the zwave device is found within a Home Assistant Addon. (Screenshot below) Would this work?

Then instead of just opening a generic port AMA* or TTY* we would be explicitly listing a device?

![Image](https://github.com/user-attachments/assets/fbc578e3-0ca5-40a3-8606-83f2ba5df692)

**Comment by alexbelgium** [comment]
Hi, could you please do a test for me?

In /addon_configs/db21ed7f_signalk/signalk.sh add the code below, then restart and paste the log here :-) thanks!
```
#!/bin/bash
apt-get 
echo "device ownership"
ls -l /dev/ttyUSB1 || true
echo "device ownership with sudo"
sudo ls -l /dev/ttyUSB1 || true
echo "try setting ownership"
for device in /dev/ttyUSB /dev/ttyUSB0 /dev/ttyUSB1; do
    if [ -e "$device" ]; then
        GROUP=$(stat -c "%G" "$device")
        sudo usermod -a -G "$GROUP" node || true
    fi
done
```

**Comment by alexbelgium** [comment]
Please try again but with USB0 instead of USB01 thanks! Also latest version uses root instead of node as user which might help (or introduce new issues)

**Comment by MrHobsons** [comment]
> Please try again but with USB0 instead of USB01 thanks! Also latest version uses root instead of node as user which might help (or introduce new issues)

After the update the Addon does not start...

Starting...
/etc/cont-init.d/00-banner.sh: executing
-----------------------------------------------------------
 Add-on: Signalk Server
 An implementation of a Signal K central server for boats
-----------------------------------------------------------
 Add-on version: 2.13.1-test3
 You are running the latest version of this add-on.
 System: Home Assistant OS 14.1  (amd64 / generic-x86-64)
 Home Assistant Core: 2025.1.3
 Home Assistant Supervisor: 2025.02.0
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums
-----------------------------------------------------------
 Provided by: https://github.com/alexbelgium/hassio-addons 
-----------------------------------------------------------
/etc/cont-init.d/00-global_var.sh: executing
/etc/cont-init.d/01-config_yaml.sh: executing
Setting permissions for the config.yaml directory
Load environment variables from /config/config.yaml if existing
If accessing the file with filebrowser it should be mapped to /addon_configs/db21ed7f_signalk/config.yaml
---------------------------------------------------------
Wiki here on how to use : github.com/alexbelgium/hassio-addons/wiki/Add‐ons-feature-:-add-env-variables
usb='true'
/etc/cont-init.d/01-custom_script.sh: executing
Execute /addon_configs/db21ed7f_signalk/signalk.sh if existing
Wiki here : github.com/alexbelgium/hassio-addons/wiki/Add-ons-feature-:-customisation
... script found, executing
[19:45:58] WARNING: This is the required permissions : crw-rw---- 1 root 18 188, 0 Feb  9 17:03 /dev/ttyUSB0
ls: cannot access '/dev/ttyUSB1': No such file or directory
Error : /etc/cont-init.d/01-custom_script.sh exiting 2

**Comment by MrHobsons** [comment]
> Hi, the only issue I see in the log is due to the usb1 in the custom script?

Hi, 
Sorry but I can't find the file  /etc/cont-init.d/01-custom_script.sh  in my system.
The error started after the update of the Addon

/anders

**Comment by alexbelgium** [comment]
Sorry I meant : if you delete the file /addon_configs/xxx-signalk/signalk.sh (or in this file change usb1 to usb0) it should start again. It is now using root so the custom script is likely not needed anymore, it could perhaps work directly!

**Comment by MrHobsons** [comment]
> Sorry I meant : if you delete the file /addon_configs/xxx-signalk/signalk.sh (or in this file change usb1 to usb0) it should start again. It is now using root so the custom script is likely not needed anymore, it could perhaps work directly!

Hi,
The update  2.13.1-2 does not work, I can start the addon but all my configuration and data is gone.
All configuration lost, Node Red and the other apps in in SignalK does not load. All Connections to NMEA2000 are lost.  Actually nothing works after the update.
The solution was to restore the backup and all works again.
Is this something you can look into?
I have SignalK on another server (no Hass) and I did update that to 2.13.2 without any problem.

**Comment by alexbelgium** [comment]
Ok thanks ; probably a question of permissions. I've reverted the code so that by default it has the same logic as before.

HOWEVER, I've put a new option "RUN_AS" in the addon options. If set to "root", it will start the app as root. I'm still ambivalent between 2 possible solutions : force the default "node" user to access the devices (but this seems restricted by HA itself), or run signalk-server as root (but then there seems to be an hidden reason why it loses access to the configuration)

Bottom line, it would be easier if I could find myself a device running on ttyUSB0 that would allow me to actually test things instead of guessing ;-)

**Comment by MrHobsons** [comment]
> Ok thanks ; probably a question of permissions. I've reverted the code so that by default it has the same logic as before.
> 
> HOWEVER, I've put a new option "RUN_AS" in the addon options. If set to "root", it will start the app as root. I'm still ambivalent between 2 possible solutions : force the default "node" user to access the devices (but this seems restricted by HA itself), or run signalk-server as root (but then there seems to be an hidden reason why it loses access to the configuration)
> 
> Bottom line, it would be easier if I could find myself a device running on ttyUSB0 that would allow me to actually test things instead of guessing ;-)

Hi,
Ok, thanks

Regarding "bottom line", since this is very important for me to solve this and just have SignalK in Home Assistant.
I wonder how we can solve this matter, what do you need exactly? I think it's easiest if you buy what you need within reasonable limits, then I can finance that purchase.

What do you think about that?

/anders

**Comment by alexbelgium** [comment]
Hi, thanks for the proposal ! I might have find a way to test it though using another USB device.

I have 2 potential solutions : could you please version test4 both directly after booting (using node user with my new permissions) ; or after enabling the RUN_AS_ROOT flag in the options then restarting.

In theory both might work, although I don't want to get your hopes too high. Thanks!

**Comment by MrHobsons** [comment]
> Hi, thanks for the proposal ! I might have find a way to test it though using another USB device.
> 
> I have 2 potential solutions : could you please version test4 both directly after booting (using node user with my new permissions) ; or after enabling the RUN_AS_ROOT flag in the options then restarting.
> 
> In theory both might work, although I don't want to get your hopes too high. Thanks!

Hi,

Great, I did update to version: 2.13.1-2_test4 and it works perfect both 
by dev/ttyUSB0 and dev/serial/by-id/usb-x
Haven´t tried with RUN_AS_ROOT and don´t think I need to?
Very pleased with this release. 
Anything else you want me to try?

/anders

**Comment by alexbelgium** [comment]
Hi great, thanks for the feedback!

I'll clean the code to remove the root related parts as they are not used anymore and release a new version

**Comment by alexbelgium** [comment]
I've pushed a final code! I'll close this then unless there is another bug

**Comment by alistairg** [comment]
This is working for me now too - thankyou!

**Comment by tingles2233** [comment]
Hi Alex, Thanks for working through this it is greatly appreciated! 

I finally had a moment in between trips to change our configuration and cabling back to HA and all is working as expected. The signalK instance now sees all the USB devices!

You are a rockstar!

Have a great Sunday!

I am good with closing up this issue/ticket.

**Comment by alexbelgium** [comment]
> Hi Alex, Thanks for working through this it is greatly appreciated!
> 
> I finally had a moment in between trips to change our configuration and cabling back to HA and all is working as expected. The signalK instance now sees all the USB devices!
> 
> I am good with closing up this issue/ticket.
> 
> [@alexbelgium](https://github.com/alexbelgium) -Sent a donation for all your hard work on this over the past year! Hope you can have a few coffee's or beers!

Thanks very much! I'll be on holiday soon perfect time for a couple nice beers !
